synapseclient/core/download/download_functions.py

```diff
@@ -719,21 +735,48 @@ def _ftp_report_hook(
                 else None
             )
 
-            response = with_retry(
-                lambda url=url, range_header=range_header, auth=auth: client._requests_session.get(
-                    url=url,
-                    headers=client._generate_headers(range_header),
-                    stream=True,
-                    allow_redirects=False,
-                    auth=auth,
-                ),
-                verbose=client.debug,
-                **STANDARD_RETRY_PARAMS,
-            )
             try:
+                response = with_retry(
+                    lambda url=url, range_header=range_header, auth=auth: client._requests_session.get(
+                        url=url,
+                        headers=client._generate_headers(range_header),
+                        stream=True,
+                        allow_redirects=False,
+                        auth=auth,
+                    ),
+                    verbose=client.debug,
+                    **STANDARD_RETRY_PARAMS,
+                )
                 exceptions._raise_for_status(response, verbose=client.debug)
             except SynapseHTTPError as err:
-                if err.response.status_code == 404:
+                if err.response.status_code == 403:
+                    url_is_expired = datetime.datetime.now(
+                        tz=datetime.timezone.utc
+                    ) + PresignedUrlProvider._TIME_BUFFER >= _pre_signed_url_expiration_time(
+                        url
+                    )
+                    if url_is_expired:
```

Comment:
Should this check also be put before:

```
                response = with_retry(
                    lambda url=url, range_header=range_header, auth=auth: client._requests_session.get(
                        url=url,
                        headers=client._generate_headers(range_header),
                        stream=True,
                        allow_redirects=False,
                        auth=auth,
                    ),
                    verbose=client.debug,
                    **STANDARD_RETRY_PARAMS,
                )
```

I am thinking of the scenario where we know before we make the call if the URL happens to be expired we can just go ahead and get a new presigned URL at that point in time.

Reply:
In one of the [tests](https://github.com/Sage-Bionetworks/synapsePythonClient/blob/f368f1fbb08b33c87443adaa37941eee219cc39a/tests/integration/synapseclient/core/test_download.py#L253-L254), `download_from_url` is called with a URL that does not have an expiration time and so this errors out if we add this check before the initial attempt. Is that a real scenario that we need to handle?

Reply:
It is a real scenario because this is what the user had reported. The way you wrote the logic is that even if we know the URL is expired we will make the call -> Fail -> Get a new URL -> make the call again.

The logic you have works, however, we could trim out the Fail and get a new token.


As for integration tests failing - You could update the logic so that if the URL does not have an expiration date/time we won't get a new token because we won't know if the failure was because it was expired, or actually denied with an HTTP 403 for some other reason.

Reply:
> URL that does not have an expiration time and so this errors out if we add this check before the initial attempt. Is that a real scenario that we need to handle?

I get what you mean about this. Let's handle this situation like this:

1. If a URL does not have an expiration time we should not error
2. Make the initial call to get the data and even if a 403 error shows up let the transfer fail and do not try again

Reply:
Implemented checking for if a URL has an expiration [here](https://github.com/Sage-Bionetworks/synapsePythonClient/blob/dc259ec854737ee7179c469b6fce6c8b221c3758/synapseclient/core/download/download_functions.py#L739-L747).